File: android/core/src/main/kotlin/org/hdhropen/kit/networking/AuthManager.kt
package org.hdhropen.kit.networking

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.hdhropen.kit.models.CurrentUser
import org.hdhropen.kit.models.UserProfile
import org.hdhropen.kit.utilities.Log

class AuthManager(
    private val apiClient: APIClient,
    private val context: Context? = null
) {
    private val _currentUser = MutableStateFlow<CurrentUser?>(null)
    val currentUser: StateFlow<CurrentUser?> = _currentUser.asStateFlow()

    private val _profiles = MutableStateFlow<List<UserProfile>>(emptyList())
    val profiles: StateFlow<List<UserProfile>> = _profiles.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _authError = MutableStateFlow<String?>(null)
    val authError: StateFlow<String?> = _authError.asStateFlow()

    private val tokenKey = "org.hdhropen.client.bearerToken"
    private val deviceIdKey = "org.hdhropen.client.deviceId"
    private val prefsName = "hdhr_open_auth_prefs"

    private val prefs: SharedPreferences? by lazy {
        context?.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
    }

    val isAuthenticated: Boolean
        get() = _currentUser.value != null

    suspend fun restoreSession() {
        prefs?.getString(deviceIdKey, null)?.let { storedDeviceId ->
            apiClient.deviceId = storedDeviceId
        }
        registerDeviceIfNeeded()

        val token = prefs?.getString(tokenKey, null) ?: return
        apiClient.bearerToken = token

        try {
            val user = apiClient.getCurrentUser()
            _currentUser.value = user
            Log.auth.info("Restored session for user: ${user.name} (${user.id})")
        } catch (e: Exception) {
            Log.auth.warning("Session restore failed, clearing token: ${e.localizedMessage}")
            prefs?.edit()?.remove(tokenKey)?.apply()
            apiClient.bearerToken = null
            _currentUser.value = null
        }
    }

    suspend fun registerDeviceIfNeeded(): Boolean {
        return try {
            val result = apiClient.registerDevice()
            prefs?.edit()?.putString(deviceIdKey, result.id)?.apply()
            apiClient.deviceId = result.id
            true
        } catch (e: Exception) {
            Log.auth.warning("Device registration failed: ${e.localizedMessage}")
            false
        }
    }

    suspend fun fetchProfiles() {
        _isLoading.value = true
        _authError.value = null
        try {
            val list = apiClient.listProfiles()
            _profiles.value = list
        } catch (e: Exception) {
            _authError.value = e.localizedMessage
            Log.auth.error("Failed to load profiles: ${e.localizedMessage}", e)
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun login(user: UserProfile, pin: String?, deviceName: String = "Android Mobile Client") {
        _isLoading.value = true
        _authError.value = null
        try {
            if (apiClient.deviceId == null) {
                registerDeviceIfNeeded()
            }

            val loggedInUser = try {
                apiClient.login(userId = user.id, pin = pin, tokenName = deviceName)
            } catch (e: APIError.Unauthorized) {
                if (e.message.contains("device", ignoreCase = true)) {
                    Log.auth.warning("Login failed due to unregistered device, auto-registering and retrying...")
                    val registered = registerDeviceIfNeeded()
                    if (registered) {
                        apiClient.login(userId = user.id, pin = pin, tokenName = deviceName)
                    } else {
                        throw e
                    }
                } else {
                    throw e
                }
            }

            loggedInUser.token?.let { token ->
                prefs?.edit()?.putString(tokenKey, token)?.apply()
                apiClient.bearerToken = token
            }
            _currentUser.value = loggedInUser
            Log.auth.info("Successfully logged in as ${loggedInUser.name}")
        } catch (e: Exception) {
            _authError.value = e.localizedMessage
            throw e
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun logout() {
        try {
            apiClient.logout()
        } catch (e: Exception) {
            Log.auth.debug("Server logout returned: ${e.localizedMessage}")
        }
        prefs?.edit()?.remove(tokenKey)?.apply()
        apiClient.bearerToken = null
        _currentUser.value = null
    }
}

File: apple/HDHROpenKit/Sources/HDHROpenKit/Networking/AuthManager.swift
import Foundation
import Security

@MainActor
public final class AuthManager: ObservableObject {
    @Published public private(set) var currentUser: CurrentUser?
    @Published public private(set) var profiles: [UserProfile] = []
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var authError: String?

    private let apiClient: APIClient
    private let tokenKey = "org.hdhropen.client.bearerToken"
    private let userIdKey = "org.hdhropen.client.userId"
    private let deviceIdKey = "org.hdhropen.client.deviceId"

    public init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    public var isAuthenticated: Bool {
        currentUser != nil
    }

    public func restoreSession() async {
        await registerDeviceIfNeeded()

        guard let token = loadKeychainString(key: tokenKey) else {
            return
        }

        await apiClient.setBearerToken(token)
        do {
            let user = try await apiClient.getCurrentUser()
            self.currentUser = user
            Log.auth.info("Restored session for user: \(user.name) (\(user.id))")
        } catch {
            Log.auth.warning("Session restore failed, clearing token: \(error.localizedDescription)")
            clearKeychain(key: tokenKey)
            await apiClient.setBearerToken(nil)
        }
    }

    /// The backend keys every login attempt off a device cookie (`get_current_device`
    /// never auto-provisions one), so a device must be registered before login can
    /// ever succeed. Idempotent server-side — safe to call on every launch and every
    /// server-address change, not just the first time.
    private func registerDeviceIfNeeded() async {
        do {
            _ = try await apiClient.registerDevice()
        } catch {
            Log.auth.warning("Device registration failed: \(error.localizedDescription)")
        }
    }

    public func fetchProfiles() async {
        isLoading = true
        authError = nil
        do {
            let list = try await apiClient.listProfiles()
            self.profiles = list
        } catch {
            authError = error.localizedDescription
            Log.auth.error("Failed to load profiles: \(error.localizedDescription)")
        }
        isLoading = false
    }

    public func login(user: UserProfile, pin: String?, deviceName: String = "Apple TV Client") async throws {
        isLoading = true
        authError = nil
        defer { isLoading = false }

        do {
            let loggedInUser = try await apiClient.login(userId: user.id, pin: pin, tokenName: deviceName)
            if let token = loggedInUser.token {
                saveKeychainString(key: tokenKey, value: token)
                await apiClient.setBearerToken(token)
            }
            self.currentUser = loggedInUser
            Log.auth.info("Successfully logged in as \(loggedInUser.name)")
        } catch {
            authError = error.localizedDescription
            throw error
        }
    }

    public func logout() async {
        do {
            try await apiClient.logout()
        } catch {
            Log.auth.debug("Server logout returned: \(error.localizedDescription)")
        }
        clearKeychain(key: tokenKey)
        await apiClient.setBearerToken(nil)
        self.currentUser = nil
    }

    // MARK: - Keychain Helpers

    private func saveKeychainString(key: String, value: String) {
        guard let data = value.data(using: .utf8) else { return }
        clearKeychain(key: key)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock
        ]
        SecItemAdd(query as CFDictionary, nil)
    }

    private func loadKeychainString(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        var dataTypeRef: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &dataTypeRef)
        guard status == errSecSuccess, let data = dataTypeRef as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private func clearKeychain(key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key
        ]
        SecItemDelete(query as CFDictionary)
    }
}
